### PR TITLE
Fail if the highlighter responds with 400 status

### PR DESCRIPTION
**Pull request by sideshowbarker**
This change causes wattsi to stop if it receives an HTTP response from the highlighter with a 400 status code.

Relates to https://github.com/whatwg/html-build/issues/201 (“Build does not fail on invalid Web IDL”)

---

This doesn’t work yet in practice for what we actually need it for, which is to make the build fail for invalid Web IDL, as reported in https://github.com/whatwg/html-build/issues/201.

The reason it doesn’t work in practice is that, as far I can tell, the highlighter doesn’t actually respond with a 400 when it encounters a Web IDL syntax error. See https://github.com/tabatkins/highlighter/issues/14.

### CODE REVIEW

File: src/wattsi.pas [hunk 1]
```diff
@@ -2207,22 +2209,36 @@ procedure Save(const Document: TDocument; const FileName: AnsiString; const InSp
...
+                     else
+                     if (AnsiContainsStr(ClassValue, 'html')) then
+                        HTTPClient.HTTPMethod('GET', HighlightServerURL + '/html?' + URLEncodedJSONContents, Ss, [200,400]);
+                     HighlighterOutput := Ss.Datastring;
+                     Ss.Free;
+                     if HTTPClient.ResponseStatusCode = 400 then
+                     begin
+                        Writeln('Error: Highlighter failed.');
+                        Write(HighlighterOutput);
+                        Halt(1);
```
Comment (domenic) [inline src/wattsi.pas]: We don't use Halt(1) anywhere else; instead we use the Fail() routine. Could we use that instead?

File: src/wattsi.pas [hunk 2]
```diff
@@ -2207,22 +2209,36 @@ procedure Save(const Document: TDocument; const FileName: AnsiString; const InSp
...
+                        HTTPClient.HTTPMethod('GET', HighlightServerURL + '/abnf?' + URLEncodedJSONContents, Ss, [200,400])
+                     else
+                     if (AnsiContainsStr(ClassValue, 'html')) then
+                        HTTPClient.HTTPMethod('GET', HighlightServerURL + '/html?' + URLEncodedJSONContents, Ss, [200,400]);
+                     HighlighterOutput := Ss.Datastring;
+                     Ss.Free;
+                     if HTTPClient.ResponseStatusCode = 400 then
+                     begin
+                        Writeln('Error: Highlighter failed.');
+                        Write(HighlighterOutput);
```
[the remark this answers is not in the document]
Comment (sideshowbarker) [reply src/wattsi.pas]:
```
                        Write(HighlighterOutput);
```
> Maybe we should Trim() here too? Doesn't matter much.

Added (better to have it)

File: src/wattsi.pas (hunk 1 above)
Comment (sideshowbarker) [reply src/wattsi.pas]: So, that `Fail()` function we have is scoped to the `ProcessDocument()` function. And the highlighting happens in the `Save()` function, so we don’t have `Fail()` there unless we either copy the code from `ProcessDocument()` function, or else move `Fail()` up to (effectively) being a global function.

In writing this patch, I had considered making that `Fail()` a global so that we could (re)use it in the `Save()` function — but then I remembered that the reason we are using `Fail()` instead of `Halt()` to begin with in the `ProcessDocument()` code is that we need all of that code to first execute all of the way through before we exit. And the reason we need that is so that, after it runs and we finally do actually exit, we need to check the exit status to see if the cause was a parser error (in which case the exit status is 65 instead of 1), so that we can re-run to get the correct line numbers (for the actual source file instead of the pre-processed source we run it on the first time).

But the `Save()` code where the highlighter is called runs after the parser has finished (and so at that point we’re processing what’s effectively the DOM representation) — and therefore there’s no reason to do that `Fail()`-indirection dance, so we can safely just do a `Halt()` to catch fire and fail right away.

Comment (domenic) [reply src/wattsi.pas]: Ah, this makes perfect sense then, thanks. I don't think we even need a comment or anything; I just missed that Fail() was scoped in that way.

File: src/wattsi.pas [hunk 3]
```diff
@@ -2207,22 +2209,36 @@ procedure Save(const Document: TDocument; const FileName: AnsiString; const InSp
...
+                        Halt(1);
+                     end;
+                     HighlighterOutputByJSONContents[CurrentHighlightedElementJSON] := HighlighterOutput;
+                     Write(F, Trim(HighlighterOutput));
+                  finally
+                     HTTPClient.Free;
+                  end;
                end;
             except
               on E: EHTTPClient do
```
[the remark this answers is not in the document]
Comment (domenic) [reply src/wattsi.pas]: I guess maybe it'd happen if a non-expected status code came back.

[the remark this answers is not in the document]
Comment (sideshowbarker) [reply src/wattsi.pas]: > I guess maybe it'd happen if a non-expected status code came back.

Right — but then the question becomes: What would cause a non-expected status code to come back? The only statuses the highlighter application code itself ever responds with are 200, 400, and 404. And it only responds with 404 if the input isn’t actually JSON, which won’t ever happen unless our code is just completely broken anyway. Outside of the highlighter application code — in the HTTP server library code — I guess there may be some cases where it responds with a 500 or other 5xx if there’s an internal failure before the highlighter application code ever gets hit. But if that happens, that’s some other bigger, more-obvious failure.

In short: I still don’t think that exception catch is ever likely to get reached in any normal cases :)
